src/unity/python/turicreate/toolkits/activity_classifier/util.py

```diff
@@ -68,7 +70,43 @@ def random_split_by_session(dataset, session_id, fraction=0.9, seed=None):
             'Input "dataset" must contain a column called %s.' % session_id)
 
     unique_sessions = _SFrame({'session': dataset[session_id].unique()})
+    if len(unique_sessions) < _MIN_NUM_SESSIONS_FOR_SPLIT:
+        print ("The dataset has less than the minimum of", _MIN_NUM_SESSIONS_FOR_SPLIT, "sessions required for train-validation split. Continuing without validation set")
+        return dataset, None
+
     chosen, not_chosen = unique_sessions.random_split(fraction, seed)
-    train = dataset.filter_by(chosen['session'], session_id)
-    valid = dataset.filter_by(not_chosen['session'], session_id)
-    return train, valid
+
+    def split_using_is_in(data, session_id, not_chosen):
+        not_chosen_filter = data[session_id].is_in(not_chosen['session'])
+
+        train = data[1 - not_chosen_filter]
+        valid = data[not_chosen_filter]
+
+        return train, valid
+
+    def split_using_filter(data, session_id, chosen, not_chosen):
+
+        # The activty classifier assumes that lines within each session
+        # are consecutive samples, ordered in ascending time order.
+        # Since filter_by() does not maintain order - we will need to re-sort
+        # after the split.
+        id_column = "original_data_row_id"
+        data = data.add_row_number(id_column)
+
+        train = data.filter_by(chosen['session'], session_id)
+        valid = data.filter_by(not_chosen['session'], session_id)
+
+        train = train.sort(id_column)
+        valid = valid.sort(id_column)
+        train = train.remove_column(id_column)
+        valid = valid.remove_column(id_column)
+
+        return train, valid
+
+    # For a small number of unique sessions, creating a filter using SFrame.is_in() would
+    # be efficient - since filtering with a binary SArray maintains order, and no sorting
+    # is required.
+    if len(unique_sessions) < _MAX_SESSIONS_TO_USE_IS_IN:
```

Comment:
For now - I've set some arbitrary hard limit, which I've found empirically.  
Giving it some more thought - the `is_in` solution takes o(k * n), where `k` is the number of sessions, and the filter+sort solution takes o(nlogn).   
Maybe I can improve this condition by setting `if number_of_sessions < log(len(dataset))`?  

@nickjong - what do you think? (that's all assuming I won't go with your suggested `apply` method).

Reply:
I think what you have is fine. There's only so far we can get with asymptotic analysis, and if this continues to be a problem, I suspect it's a better investment of our time to find a way to make the O(n) approach work.